src/julia-syntax.scm

```diff
@@ -2738,15 +2752,28 @@
                               (vinfo:set-sa! vi #t))
                           (vinfo:set-asgn! vi #t)))
                e)
-             (begin (analyze-vars (caddr e) env captvars sp)
+             (begin (analyze-vars (caddr e) env captvars sp glob-assign)
                     (assert (eq? (car (cadddr e)) 'lambda))
                     (analyze-vars-lambda (cadddr e) env captvars sp
-                                         (method-expr-static-parameters e)))))
+                                         (method-expr-static-parameters e)
+                                         glob-assign))))
         ((module toplevel) e)
-        (else (for-each (lambda (x) (analyze-vars x env captvars sp))
+        (else (for-each (lambda (x) (analyze-vars x env captvars sp glob-assign))
                         (cdr e))))))
 
-(define (analyze-variables! e) (analyze-vars e '() '() '()) e)
+(define (analyze-variables! e)
+  (let ((glob-assign (table)))
+    (analyze-vars e '() '() '() glob-assign)
+    ;; if we collected any assignments to globals
+    ;; annotate them now at the toplevel
+    (let ((glob-decl (map (lambda (e) `(global ,e)) (table.keys glob-assign))))
+      (if (null? glob-decl)
+          e
+          (insert-after-meta
+            (if (and (pair? e) (eq? (car e) 'block))
+                e
+                `(block ,e))
+            glob-decl)))))
```

Comment:
This moves the global decls from local scope (where the syntax does not inherently create new bindings), to global scope (where it does). I'm generally inclined to do all scope and syntax lowering here, rather than implementing any part of it in post-processing code.